src/compiler/checker.ts

```diff
@@ -8964,6 +8964,13 @@ namespace ts {
                         (isTypeSubsetOf(globalObjectType, target) || (!isComparingJsxAttributes && isEmptyObjectType(target)))) {
                         return false;
                     }
+                    if (target.flags & TypeFlags.Union) {
+                        const discriminantType = findMatchingDiscriminantType(source, target as UnionType);
+                        if (discriminantType)  {
+                            // check excess properties against discriminant type only, not the entire union
+                            return hasExcessProperties(source, discriminantType, reportErrors);
+                        }
+                    }

```

Comment:
I'm not sure it's that simple. What if multiple types in the target have a matching discriminant? I think there's an implicit assumption in your code that each target constituent has a unique discriminant, which may not be the case.